en/dev_log/log_encryption.md

```diff
@@ -0,0 +1,179 @@
+You can tell which one is the symmetric key by the file size.
+The symmetric key is normally pretty small (around 300 bytes).
+:::
+
+The process for downloading log files is the same, except you will download two separate files per log.
+Each time you fly, a log file, and a symmetric key are generated.
+The only way to tell them apart is by file size.
+
+As normal, go to **Analyze Tools -> Log Download** and select the log you wish to download.
+
+Make sure to download both the encrypted log and symmetric key that were generated.
+
+![QGroundControl ULog Download](../../assets/qgc/analyze/encrypted_log.png)
+
+This photo shows both a generated symmetric key and the associated encrypted ULog file.
+
+### Decrypt ULogs
+
+Before you can analyze your encrypted logs, you will need to decrypt them.
+There is a Python script that can be used to decrypt logs in Tools/decrypt_ulog.py.
+
+decrypt_ulog.py takes 3 arguments:
+
+1. The encrypted `.ulogc` file.
+2. The symmetric key `.ulogk` file.
+3. The decryption key (the RSA2048 `.pem` private key which is used to unwrap the `.ulogk` file).
+
+```sh
+usage: decrypt_ulog.py [-h] [ulog_file] [ulog_key] [rsa_key]
+
+CLI tool to decrypt an ulog file
+
+positional arguments:
+  ulog_file   .ulog file
+  ulog_key    .ulogk, encrypted key
+  rsa_key     .pem format key for decrypting the ulog key
+
+optional arguments:
+  -h, --help  show this help message and exit
+
+```
+
+As an example:
+
+```sh
+python3 decrypt_ulog.py \
+/home/john/Downloads/log_24_2024-10-6-23-39-50.ulg \
+/home/john/Downloads/log_23_2024-10-6-23-39-48.ulg \
+new_keys/private_key.pem
+```
+
+There will be no printed output on a successful decryption, and a new file is created with the ".ul" suffix instead of ".ulg".
+Rename this back to a .ulg file and it is now ready for flight review!
+
+### Generate Your Own Keys
+
+In a production environment, you should your own generated keys rather than using the example keys.
+
+- To generate the key0 public key, you can run the `Tools/cryptotools.py` python program.
+
+As an example, this generates a file called new.pub and a new.json file in Tools/test_keys:
+
+```sh
+python3 Tools/cryptotools.py  --genkey Tools/test_keys/new
+```
+
+- To generate a rsa2048 private and public key, you can use openssl
+
+```sh
+openssl genpkey -algorithm RSA -out private_key.pem -pkeyopt rsa_keygen_bits:2048
+```
+
+Then, you can create a public key from this private key,
+
+```sh
+# Convert private_key.pem to a DER file
+openssl rsa -pubout -in private_key.pem -outform DER -out public_key.der
+# From the DER file, generate a public key in hex format, seperated by commas
+xxd -p public_key.der | tr -d '\n' | sed 's/\(..\)/0x\1, /g' > public_key.pub
+```
+
+Now, you would modify your `.px4board` file to point CONFIG_PUBLIC_KEY0 to the path of `new.pub` that was generated via `cryptotools.py`, and set CONFIG_PUBLIC_KEY1 to `rsa2048.pub`.
+The private key generated can be used in the decryption environment.
```

Comment:
So again, not clear what the two keys are for - why you need both. Also a bit confused about this  new.json file here.

Can you provide a real example of the kconfig values for these keys? E.g. screenshot in kconfig menu? 
Are there any restrictions on the locations? I assume you must build the keys somewhere under the PX4 root, for example?

Reply:
@hamishwillee 

I did some digging into PX4 source code to find an answer to this. It makes sense to me now.

The current implementation of Crypto contains a keystore that can hold 4 separate keys. 
These are contained in CONFIG_PUBLIC_KEY0, CONFIG_PUBLIC_KEY1, CONFIG_PUBLIC_KEY2, CONFIG_PUBLIC_KEY3

Two params exist in PX4

[SDLOG_EXCH_KEY](https://docs.px4.io/main/en/advanced_config/parameter_reference.html#SDLOG_EXCH_KEY), and [SDLOG_KEY](https://docs.px4.io/main/en/advanced_config/parameter_reference.html#SDLOG_KEY). 

By default, SDLOG_EXCH_KEY is set to 1, and SDLOG_KEY is set to 2.

So by default, CONFIG_PUBLIC_KEY0 is our exchange key, and CONFIG_PUBLIC_KEY1 is our encryption key.

_______________
_**Note of a concern I have about these parameters:**_
 SDLOG_EXCH_KEY and SDLOG_KEY defaults are 1 and 2. Yet the parameter itself has a minimum value of 0. And when I'm looking in the source code, I cannot find anywhere that these param values are getting subtracted by 1 when indexing the keystore. I am wondering if there is an unintended bug here when running the default cryptotest example.
I am going to do more digging here
___________________
 

(take this with a grain of salt because cryptography is not my area of expertise) I believe that the exchange key exists so that the encryption key can be stored in an encrypted format on the SD card.


> Can you provide a real example of the kconfig values for these keys? E.g. screenshot in kconfig menu?

I don't think kconfig has an area to change these keys.

> Are there any restrictions on the locations? I assume you must build the keys somewhere under the PX4 root, for example?

Not that I can tell. As a test, I placed key1 in my home directory and changed the .px4board file and it built just fine.

Reply:
@jnomikos Thanks! Looks like we have had some crossover in the last hour. Let me digest this.

Reply:
OK, so I think I get what you are saying.
- Keystore has up to 4 places for keys CONFIG_PUBLIC_KEY0, CONFIG_PUBLIC_KEY1, CONFIG_PUBLIC_KEY2, CONFIG_PUBLIC_KEY3
- We store, `new.pub` in `CONFIG_PUBLIC_KEY0` and our rsapub key in  `CONFIG_PUBLIC_KEY1`. We could also store a notional `new2.pub` in  `CONFIG_PUBLIC_KEY2` and rsapub2 key in  `CONFIG_PUBLIC_KEY3`.
- The  [SDLOG_EXCH_KEY](https://docs.px4.io/main/en/advanced_config/parameter_reference.html#SDLOG_EXCH_KEY), and [SDLOG_KEY](https://docs.px4.io/main/en/advanced_config/parameter_reference.html#SDLOG_KEY) are used to select the _index_ of the keys in the store that are supposed to be used in calculations.
- SDLOG_EXCH_KEY defaults to 1 and is supposed to store the location of the public key that is used to encypt the symmetric key.
  - It is not clear if this is indexed against the keystore or the set of log keys (?) - i.e. the two possible keys. Should be stated in the docs.
- SDLOG_KEY default to 2 and is supposed to store the location of the symmetric key that is generated.
  - It is not clear if this is indexed against the keystore or the set of log keys (?)
- Assuming zero index key store and that the params above are both relative to the key store, that means
  - SDLOG_KEY (symmetric key)=2=CONFIG_PUBLIC_KEY3= not set by us, which is fine.
  - SDLOG_EXCH_KEY (rsa key)=1=CONFIG_PUBLIC_KEY1= our RSA key
  - CONFIG_PUBLIC_KEY0 and CONFIG_PUBLIC_KEY2 are not used.
  - (new.pub) is not used
  
I think this will work because from what I can tell `new.pub` serves no purpose. My assumption is that as long as our RSA key is mapped to the correct key that we set in the config, and _nothing else is trying to write to `CONFIG_PUBLIC_KEY3`_ which we are using for accessing the symmetric key, we should be fine.

It would be great to confirm so that we can remove setting and using new.pub and that script.